skbio/metadata/_interval.py

```diff
@@ -0,0 +1,274 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from ._intersection import IntervalTree
+
+
+class Interval():
+    '''Store the metadata of a sequence interval.
+
+    Parameters
+    ----------
+    intervals : list of tuple of ints
+        List of tuples representing start and end coordinates.
+    boundaries : list of tuple of bool
+        List of tuples, representing the openness of each interval.
+    metadata : dict
+        Dictionary of attributes storing information of the feature
+        such as `strand`, `gene_name` or `product`.
+    _interval_metadata : object
+        A reference to the `IntervalMetadata` object that this
+        interval is associated to.
+    '''
+    def __init__(self, intervals=None, boundaries=None,
+                 metadata=None, _interval_metadata=None):
+        self._intervals = []
+        for interval in intervals:
+            inv = _polish_interval(interval)
+            self._intervals.append(inv)
+        self.boundaries = boundaries
+        self.metadata = metadata
+        self._interval_metadata = _interval_metadata
+
+    def __getitem__(self, key):
+        return self.metadata[key]
+
+    def __setitem__(self, key, val):
+        self.metadata[key] = val
+
+    def __lt__(self, other):
+        return self.intervals < other.intervals
+
+    def __gt__(self, other):
+        return self.intervals > other.intervals
+
+    def __le__(self, other):
+        return self.intervals <= other.intervals
+
+    def __ge__(self, other):
+        return self.intervals >= other.intervals
+
+    def __eq__(self, other):
+        return ((self.metadata == other.metadata) and
+                (self.intervals == other.intervals) and
+                (self.boundaries == other.boundaries))
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __repr__(self):
+        return ''.join([
+            "Interval("
+            "intervals=" + str(self.intervals),
+            ", metadata=" + str(self.metadata),
+            ")"])
+
```

Comment:
It is also important in that it ensures an `Interval` object is capable of existing in a dropped state. Suppose I did this:
```python
intervals = list(interval_metadata.query(**some_args))
interval_metadata.drop(**some_args)
# What are `intervals` now if they don't support being dropped themselves?
```
So while `Interval.drop` is itself convenient, it also ensures it is "droppable" (much like a `close` on a filehandle).

There may also be situations where you are trying to do some kind of inclusion exclusion logic, where you might perform several queries and build sets where you intend to drop the results after unioning/intersecting a bunch.

Related, `Interval.drop` should be idempotent, in that calling `drop` twice has the same effect as calling it once. Especially in a union situation, some Intervals might show up more than once (granted calling `IntervalMetadata.drop` several times is the more effective approach here, but the API needs to work regardless).